### PR TITLE
#fLTIEPWT Assign buses to routes

### PR DESCRIPTION
**Pull request by Ally4**
#### What does this PR do?

- Assign buses to routes

#### Description of Task to be completed?

- Installation of sequelize ORM

- Creation models

- Creation of migrations

- Creation of the controller of assigning buses to route

- Creation of the controller of getting the buses assigned to routes in pagination

- Creation of validation the controller of assigning buses to route

- Creation of router of assigning buses to route

- Creation of router of getting assigning buses to route

#### How should this be manually tested?

- By cloning this repository

- By running `npm install` in the terminal

- By running `npm run dev` in the terminal

- By running `npm run migrate:reset` in the terminal

- By running `npm run seed` in the terminal

- The available routes are: `304` and `305`

- By opening on this documentation: `http://localhost:3020/api-docs/#/`

- Running signing in as operator with this credentials: `email:el.ally741operator@gmail.com` and `password:admin`

- By putting the `token` from the signin in the token place

#### Any background context you want to provide?

- Installation of a new package

#### What are the relevant pivotal tracker stories?

- `https://trello.com/c/fLTlEPWT/13-assign-buses-to-routes`

#### Screenshots (if appropriate)

Screen Shot 2020-11-18 at 10 21 01
Screen Shot 2020-11-18 at 10 21 13
Screen Shot 2020-11-18 at 10 21 40
Screen Shot 2020-11-18 at 10 21 48
Screen Shot 2020-11-18 at 10 22 00
Screen Shot 2020-11-18 at 10 22 19
Screen Shot 2020-11-18 at 10 22 23

### CODE REVIEW

**Review by TuyizeeAnastase — APPROVED** [review]

**Review by Ntwali-Josue — APPROVED** [review]

**Review by Mugema123 — APPROVED** [review]

File: src/controllers/assignBusesToRoutes.js [hunk 1]
```diff
@@ -0,0 +1,98 @@
...
+      });
+
+      if (!findRoute) {
+        return res.status(404).json({
+          status: 404,
+          message: res.__('The route you are looking for is not in the system'),
+        });
+      }
+
+      const toUpdate = await Buses.findOne({
```
Comment (ceelogre) [inline src/controllers/assignBusesToRoutes.js]: Use descriptive names.

File: src/controllers/assignBusesToRoutes.js [hunk 2]
```diff
@@ -0,0 +1,98 @@
...
+            as: 'route',
+            attributes: ['id', 'direction', 'routeID'],
+          }],
+        });
+        return res.status(200).json({
+          status: 200,
+          message: 'the bus was assigned successfully',
+          bus: assignedBus,
+        });
+      }
```
Comment (ceelogre) [inline src/controllers/assignBusesToRoutes.js]: This is a long block for one function. Consider refactoring to move some of the functionalities to utilities.

File: src/database/models/buses.js [hunk 3]
```diff
@@ -0,0 +1,17 @@
+module.exports = (sequelize, DataTypes) => {
```
Comment (ceelogre) [inline src/database/models/buses.js:3]: We are using ES Modules so `module.exports` is not allowed.

File: src/test/bus.test.js [hunk 4]
```diff
@@ -62,9 +62,9 @@ describe('Buses', () => {
...
     chai
       .request(app)
-      .get('/api/v1/buses/2')
+      .get('/api/v1/buses/100')
       .set('x-access-token', adminToken)
-      .set('request.params', 2)
+      .set('request.params', 100)
       .end((err, res) => {
         expect(res).to.have.status(401);
         expect(res).to.have.property('status');
```
Comment (ceelogre) [inline src/test/bus.test.js:70]: The second assertion is extraneous. Also, it'd be better to add another assertion.

File: .env [hunk 5]
```diff
@@ -1,6 +1,6 @@
 DATABASE_URL= 
-DEV_DATABASE_URL=postgres://postgres:phantom@127.0.0.1:5432/users
-TEST_DATABASE_URL=postgres://postgres:phantom@127.0.0.1:5432/test_db
+DEV_DATABASE_URL=postgres://postgres:ally@127.0.0.1:5432/development
+TEST_DATABASE_URL=postgres://postgres:ally@127.0.0.1:5432/test
 ACCESS_TOKEN_SECRET = swsh23hjddnnsij
 ACCESS_TOKEN_LIFE = 7days
 PORT = 3020
```
Comment (ceelogre) [inline .env]: Why did you push `.env` ?

**Review by ceelogre — CHANGES_REQUESTED** [review]
You might wanto to add `type:module` to your package.json to resolve most of the hound violations.
Also, your messages are not translated.
Capture d’écran 2020-11-18 à 10 42 01
Capture d’écran 2020-11-18 à 10 43 07

File: .env (hunk 5 above)
Comment (Ally4) [reply .env]: going to refresh, as its in .gitignore

Comment (ceelogre) [reply .env]: This is the commit that introduced it: https://github.com/atlp-rwanda/eagles-phantom-backend/commit/0f451a4b323516784be39bb256d6d8655423af05

You probably just rebased.

File: src/controllers/assignBusesToRoutes.js (hunk 2 above)
Comment (Ally4) [reply src/controllers/assignBusesToRoutes.js]: worked on

File: .sequelizerc [hunk 6]
```diff
@@ -1,7 +1,7 @@
 const path = require('path')
 
 module.exports = {
-  config: path.resolve('./src/database/config', 'config.js'),
+  config: path.resolve('./src//database/config', 'config.js'),
```
Comment (blaise82) [inline .sequelizerc]: why did you put two //
use a correct path (don't change it)

File: .travis.yml [hunk 7]
```diff
@@ -1,10 +1,9 @@
 services:
   - postgresql
 before_script:
-  - psql -c 'CREATE DATABASE users;' -U postgres
-  - psql -c 'CREATE DATABASE test_db;' -U postgres
-
-install: npm install 
+  - psql -c 'CREATE DATABASE test;' -U postgres
+  - psql -c 'CREATE DATABASE development;' -U postgres
```
Comment (blaise82) [inline .travis.yml]: remove the development database from travis

File: package.json [hunk 8]
```diff
@@ -20,8 +20,9 @@
   },
   "nyc": {
     "exclude": [
-      "src/database/migrations",
+      "src/database/migrations/",
       "src/database/seeders",
+      "src/database/models",
```
Comment (blaise82) [inline package.json]: u can just include `src/database`

File: src/controllers/assignBusesToRoutes.js [hunk 9]
```diff
@@ -0,0 +1,83 @@
...
+        return AssignBusesToRoutesResponses(res, 404, 'The route you are looking for is not in the system');
+      }
+
+      const busToAssign = await Buses.findOne({
+        where: { id: req.params.id },
+      });
+
+      if (!busToAssign) {
+        return AssignBusesToRoutesResponses(res, 404, 'The bus you are looking for is not in the system');
+      }
```
Comment (blaise82) [inline src/controllers/assignBusesToRoutes.js]: create two middlewares 
one for finding the route and another one for finding the bus (if they are not there yet)

File: src/controllers/assignBusesToRoutes.js [hunk 10]
```diff
@@ -0,0 +1,83 @@
...
+        include: [{
+          model: Routes,
+          as: 'route',
+          attributes: ['id', 'direction', 'routeID'],
+        }],
+      });
+      const pagination = paginate(page, count, rows, limit);
+
+      if (offset >= count) {
+        return AssignBusesToRoutesResponses(res, 404, 'The page you are looking for is not yet created');
```
Comment (blaise82) [inline src/controllers/assignBusesToRoutes.js]: provide a better message

File: src/database/migrations/20201111144427-create-buses.js [hunk 11]
```diff
@@ -0,0 +1,38 @@
...
+      updatedAt: {
+        allowNull: false,
+        type: Sequelize.DATE,
+      },
+    });
+  },
+  down: async (queryInterface) => {
+    await queryInterface.dropTable('Buses');
+  },
+};
```
Comment (blaise82) [inline src/database/migrations/20201111144427-create-buses.js]: have one bus create a migration 
and if you need to update the bus table do create a new update migration

**Review by blaise82 — DISMISSED** [review]

File: src/database/models/buses.js (hunk 3 above)
Comment (ceelogre) [reply src/database/models/buses.js:3]: This is not addressed up to this moment.

**Review by ceelogre — APPROVED** [review]

**Pull request merged**